common/managed_thread.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <cassert>
+#include <thread>
+#include <functional>
+#include <atomic>
+
+namespace ILLIXR {
+
+/**
+ * @brief An object that manages a std::thread; it joins and exits when the object gets destructed.
+ */
+class managed_thread {
+private:
+	std::atomic<bool> _m_stop {false};
+	std::thread _m_thread;
+	std::function<void()> _m_body;
+	std::function<void()> _m_on_start;
+	std::function<void()> _m_on_stop;
+
+	void thread_main() {
+		assert(_m_body);
```

Comment:
We have a bunch of assertions throughout the code. Is our intention to have these definitely be optimized out during compilation with the `opt` profile? Not an issue, just curious. If we want assertions to be more readable, we can also use the old string literal trick: `assert(bool_condition && "Assertion failed message")`.